Additions to utils.data: get_type(), glob_list(), list_rm_match(), replace_list_element(), updated_dict()

gluf-technik opened the pull request:
### What does this PR do?

Adds some functions to `salt/utils/data.py`
  - `get_type()` to query a Python type inside a Jinja template
  - `glob_list()` to get only the elements of a list matching the glob
  - `list_rm_match()` to get a news list without the elements matching a regex
  - `replace_list_element()` to turn `[a, bx, c]` into `[a, b1, b2 ,b3, c]`
  - `updated_dict()` to get an update of a dict without modifying the original one

I think all five are results of trying to make Jinja do things it's not made for
and at lease some of them may even be suitable for becoming Jinja filters.

Pointers to other modules doing those things are welcome, too, so we don't
have to maintain custom modules until this PR gets merged, even if only
partially ; )

### What issues does this PR fix or reference?
Fixes: `None`

### Merge requirements satisfied?
**[NOTICE] Bug fixes or features added to Salt require tests.**
<!-- Please review the [test documentation](https://docs.saltproject.io/en/master/topics/tutorials/writing_tests.html) for details on how to implement tests into Salt's test suite. -->
- [ ] Docs
- [ ] Changelog - https://docs.saltproject.io/en/master/topics/development/changelog.html
- [ ] Tests written/updated

### Commits signed with GPG?
~~Yes~~/No

Please review [Salt's Contributing Guide](https://docs.saltproject.io/en/master/topics/development/contributing.html) for best practices.

See GitHub's [page on GPG signing](https://help.github.com/articles/signing-commits-using-gpg/) for more information about signing commits with GPG.

gluf-technik committed:
Add replace_list_element() to utils/data.py

Takes a list `orig` and inserts the elements of `updates_list`
anyplace where an element equal `placeholder` has been.
In list ['a', 'bx', 'c'] placeholder 'bx' would be replaced
with the elements of ['b1', 'b2', 'b3'] resulting in list
['a', 'b1', 'b2', 'b3', 'c'].

TODO: Update docs

gluf-technik committed:
Add glob_list() to utils/data.py

Uses fnmatch for "globbing" elements of a list:
The globbing the list [eth0, eth2, lo] with pattern
'eth*' results in [eth0, eth2]

Same function as used in Salt's match.glob.
https://docs.python.org/3/library/fnmatch.html

TODO:
  - Update docs
  - Add `jinja_filter("glob_list")`?

gluf-technik committed:
Add list_rm_match() to utils/data.py

Returns a list with those elements of the
argument `listing`  which __do not__  match
the regular expression `rgx`.

TODO:
  - Update docs
  - Add `jinja_filter("list_rm_match")`?

gluf-technik committed:
Add get_type() to check Python types in Jinja.

Very useful for debugging templates.

TODO:
  - Update docs
  - Add `jinja_filter("get_type")`

gluf-technik committed:
updated_dict() so copying a dict & .update() isn't neccessary

AFAIK it's not possible in Jinja to get a new, updated/merged
dictionary, just using `{% do ...%}` with `.update()` to update
one inplace.

TODO:
  - Update docs
  - Add `@jinja_filter("update")`?

gluf-technik committed:
Trying to keep things in alpabetical order

dwoz reviewed (changes requested):
@gluf-technik This needs to be re-based, tests added, and a changelog entry.

dwoz committed:
Fix code issues in data.py utility additions

- list_rm_match: use compiled regex so ignorecase/multiline flags are
  actually applied (they were built but never passed to re.match)
- list_rm_match: fix unbalanced backtick in log.debug format string
- replace_list_element: fix typo 'Neiter' -> 'Neither'
- replace_list_element: restructure type-checks to avoid possible
  UnboundLocalError on the msg variable
- All five functions: replace bare TypeError with SaltInvocationError
- All five functions: expand docstrings with versionadded, parameter
  docs, and CLI/Jinja examples
- get_type: return type.__name__ (e.g. 'int') instead of the repr
  string '<class \'int\'>' which is more useful in Jinja templates

dwoz committed:
Add tests for data.py utility additions; fix replace_list_element algorithm

Add 30 unit tests covering get_type(), glob_list(), list_rm_match(),
replace_list_element(), and updated_dict().

Also fix a logic bug in replace_list_element(): the reverse/insert
approach produced incorrect element ordering when the placeholder
appears multiple times. Replace with a simple forward extend().

dwoz committed:
Add changelog entry for PR #64703
